lib/netsuite/async_result.rb

```diff
@@ -0,0 +1,59 @@
+# https://system.netsuite.com/help/helpcenter/en_US/Output/Help/SuiteCloudCustomizationScriptingWebServices/SuiteTalkWebServices/AsynchronousRequestProcessing.html
+module NetSuite
+  class AsyncResult
+
+    attr_reader :errors, :type
+
+    def initialize(async_result)
+      @type = async_result[:"@xsi:type"]
+
+      base_ref_list = []
+      errors = []
+
+      if async_result[:read_response_list]
+        response_list = async_result[:read_response_list][:read_response]
+      else
+        response_list = async_result[:write_response_list][:write_response]
+      end
+
+      Array[response_list].flatten.each do |response|
+        if response[:status][:@is_success] == 'true'
+          base_ref_list << response[:base_ref]
+        elsif response[:status][:status_detail]
+           response_errors = Array[response[:status][:status_detail]].flatten.map do |details|
+             NetSuite::Error.new(details)
+           end
+           errors << [response[:base_ref][:@external_id], response_errors]
+        end
+      end
+
+      @base_ref_list = NetSuite::Records::BaseRefList.new(base_ref: base_ref_list)
+      @errors = Hash[errors]
```

Comment:
How about `@errors = {}`...  `@errors[response[:base_ref][:@external_id]]` = response_errors

This way you can get rid of `error = []`